File: src/core/net/strategies/direct.ts
import type { FetchContext, FetchStrategy, NgaRequest, StrategyOutcome } from '../types'
import { runAttempt } from './attempt'

export const DIRECT_STRATEGY_NAME = 'direct'

/**
 * 反封锁链的第一档：直连官方域名发**一次**请求。
 *
 * 只发一次是它与 format-rotation 的全部区别：不轮换、不换账号、不碰缓存。
 * 单测与不需要反封锁的场合（联调、smoke）用它，设备侧的链用 format-rotation。
 *
 * 失败分类决定链要不要往下走：解析失败/HTTP 状态错误 ≈ 被封（可重试），
 * 服务端明确的语义错误（找不到主题之类）不重试，直接抛给调用方。
 */
export function createDirectStrategy(): FetchStrategy {
  return {
    name: DIRECT_STRATEGY_NAME,
    run(request: NgaRequest, context: FetchContext): Promise<StrategyOutcome> {
      return runAttempt(request, context, {
        via: DIRECT_STRATEGY_NAME,
        combo: {
          format: request.format ?? 'json',
          host: request.host ?? context.host,
        },
      })
    },
  }
}

File: native/app/src/main/kotlin/com/chasel/ng2n/core/net/strategies/Direct.kt
package com.chasel.ng2n.core.net.strategies

import com.chasel.ng2n.core.net.FetchCombo
import com.chasel.ng2n.core.net.FetchContext
import com.chasel.ng2n.core.net.FetchStrategy
import com.chasel.ng2n.core.net.NgaRequest
import com.chasel.ng2n.core.net.ResponseFormat
import com.chasel.ng2n.core.net.StrategyOutcome

const val DIRECT_STRATEGY_NAME = "direct"

/**
 * 直连:按当前域名发**一次**请求。直译 `src/core/net/strategies/direct.ts`。
 *
 * 只发一次是它与 format-rotation 的全部区别:不轮换、不换账号、不碰缓存。
 *
 * **写操作的链就只有这一档**(修 P1-01):服务端可能已经执行了、只是响应丢在路上,
 * 重发会产生重复收藏夹 / 把点赞翻回去,换账号更会把写入落到别人头上。
 * 失败即终点,由调用方决定要不要让用户重来。
 */
class DirectStrategy : FetchStrategy {

  override val name: String = DIRECT_STRATEGY_NAME

  override suspend fun run(request: NgaRequest, context: FetchContext): StrategyOutcome =
    runAttempt(
      request = request,
      context = context,
      via = DIRECT_STRATEGY_NAME,
      combo = FetchCombo(
        format = request.format ?: ResponseFormat.JSON,
        host = request.host ?: context.host,
      ),
    )
}
